### ISSUE TITLE
C++: Add operator+() to DataContainer

### ISSUE THREAD

**Issue by KrisThielemans** [SyneRBI/SIRF#997]
We have `add()` etc, but it's more natural to have just use `operator+`. Also, if we use SWIG, it will automatically generate Python operators.

I'd just create inline functions that call the current `add` etc.

**Comment by evgueni-ovtchinnikov** [comment]
@KrisThielemans so far as I understand, the result of `operator+` is returned via the copy constructor and so involves wasteful temporary duplication of data (3 of them for computing `a*x + b*y`)

**Comment by evgueni-ovtchinnikov** [comment]
@KrisThielemans it looks like duplication can be avoided by using `std::move` - studying...

**Comment by KrisThielemans** [comment]
yes, there's also the (much older) https://en.wikipedia.org/wiki/Copy_elision#Return_value_optimization.

However, I believe one of the complications here is the return value. Obviously, we cannot return a `DataContainer`  as that's abstract. We could return `DataContainer&` (which I've seen on the internet) but that seems unsafe/not compilable (a reference to a temp object? hmm.). I can't see how `std::move` helps with this.

What should be easy is to implement `a += b`. No temp copy involved.

If you want to study, here's a fun way to avoid lots of overhead in a*x + b*y: [Expression templates](https://devtut.github.io/cpp/expression-templates.html#basic-expression-templates-on-element-wise-algebraic-expressions)
